### PR TITLE
Added deploy input validation

### PR DESCRIPTION
**Pull request by Radinyn**

### CODE REVIEW

File: protostar/starknet_gateway/gateway_facade_test.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        0xC,  # inst_list_arg[0].c
+        0x1,  # inst_list_arg[1].a
+        0x2,  # inst_list_arg[1].b
+        0x3,  # inst_list_arg[1].c
+    ]
+
+    validate_deploy_input(COMPILED_CONTRACT_ABI_ONLY, inputs)
+
+
+def test_validating_deploy_inputs_not_enough():
```
Comment (mkaput) [inline protostar/starknet_gateway/gateway_facade_test.py:153]: I would also add a test for empty inputs list.

File: protostar/starknet_gateway/gateway_facade.py [hunk 2]
```diff
@@ -281,3 +284,44 @@ def __init__(self, function_name: str):
 class ContractNotFoundException(ProtostarException):
     def __init__(self, contract_address: AddressRepresentation):
         super().__init__(f"Tried to call unknown contract:\n{contract_address}")
+
+
+class InvalidInputException(ProtostarException):
+    pass
+
+
+def validate_deploy_input(compiled_contract_json: str, inputs: List[int]) -> None:
```
Comment (mkaput) [inline protostar/starknet_gateway/gateway_facade.py:293]: can you move logic of json parsing out of this function?

File: protostar/starknet_gateway/gateway_facade.py [hunk 3]
```diff
@@ -281,3 +284,44 @@ def __init__(self, function_name: str):
...
+            return 1
+
+        def type_filter(element: Dict):
+            return element["name"] == typename
+
+        [type_definition] = filter(type_filter, abi)
+        return type_definition["size"]
+
+    abi = json.loads(compiled_contract_json)["abi"]
+    [constructor] = filter(constructor_filter, abi)
```
Comment (mkaput) [inline protostar/starknet_gateway/gateway_facade.py:308]: 1. List comprehension would be more pythonic here.
2. Is there a possibility that there are two constructors in ABI?

File: protostar/starknet_gateway/gateway_facade.py [hunk 4]
```diff
@@ -281,3 +284,44 @@ def __init__(self, function_name: str):
...
+    expected_inputs = constructor["inputs"]
+
+    i = 0
+    for expected_input in expected_inputs:
+        raw_type_size = get_type_size(abi, expected_input["type"].rstrip("*"))
+        if expected_input["type"].endswith("*"):
+            size = inputs[i - 1] * raw_type_size
+        else:
+            size = raw_type_size
```
Comment (mkaput) [inline protostar/starknet_gateway/gateway_facade.py:317]: Could this code use `parse_type` function from starkware/cairo/lang/compiler/parser.py ?

File: protostar/starknet_gateway/gateway_facade.py [hunk 5]
```diff
@@ -281,3 +284,44 @@ def __init__(self, function_name: str):
...
+
+class InvalidInputException(ProtostarException):
+    pass
+
+
+def validate_deploy_input(compiled_contract_json: str, inputs: List[int]) -> None:
+    def constructor_filter(element: Dict):
+        return element["type"] == "constructor"
+
+    def get_type_size(abi: List[Dict], typename: str):
```
Comment (mkaput) [inline protostar/starknet_gateway/gateway_facade.py:297]:
```suggestion
    def get_type_size(abi: AbiType, typename: str):
```

Comment (mkaput) [inline protostar/starknet_gateway/gateway_facade.py:297]: This function could be put as top-level one.

**Review by mkaput — CHANGES_REQUESTED** [review]

File: protostar/starknet_gateway/gateway_facade.py (hunk 2 above)
Comment (Radinyn) [reply protostar/starknet_gateway/gateway_facade.py:293]: we should probably try-catch it too

File: protostar/starknet_gateway/gateway_facade.py (hunk 4 above)
Comment (Radinyn) [reply protostar/starknet_gateway/gateway_facade.py:317]: I don't think it gives us what we need

File: protostar/starknet_gateway/gateway_facade_test.py (hunk 1 above)
Comment (Radinyn) [reply protostar/starknet_gateway/gateway_facade_test.py:171]: done

File: protostar/starknet_gateway/gateway_facade.py (hunk 2 above)
Comment (Radinyn) [reply protostar/starknet_gateway/gateway_facade.py]: done

File: protostar/starknet_gateway/gateway_facade.py (hunk 3 above)
Comment (Radinyn) [reply protostar/starknet_gateway/gateway_facade.py]: 1. done

File: protostar/starknet_gateway/gateway_facade.py (hunk 5 above)
Comment (Radinyn) [reply protostar/starknet_gateway/gateway_facade.py]: done

Comment (Radinyn) [reply protostar/starknet_gateway/gateway_facade.py]: done

File: protostar/starknet_gateway/gateway_facade.py (hunk 2 above)
Comment (Radinyn) [reply protostar/starknet_gateway/gateway_facade.py]: > we should probably try-catch it too

on the other hand I don't think checking for invalid input is a job of this module, wdyt?

File: protostar/starknet_gateway/gateway_facade.py (hunk 4 above)
Comment (mkaput) [reply protostar/starknet_gateway/gateway_facade.py]: it will give you information if you have raw `felt` or a pointer to something. this will definitely be more robust and future proof (what happens if starkware changes pointer syntax?) then manual parsing

File: protostar/starknet_gateway/gateway_facade.py (hunk 2 above)
Comment (mkaput) [reply protostar/starknet_gateway/gateway_facade.py]:
```suggestion
def validate_deploy_input(abi: AbiType, inputs: List[int]) -> None:
```

This is the interface I would expect here. Currently, this function is mixing concerns: parsing and contract input validation (actually the latter could be divided into two stages -- finding the constructor and then validating its inputs, but this is less important)

File: protostar/starknet_gateway/gateway_facade_input_validation_test.py [hunk 6]
```diff
@@ -0,0 +1,284 @@
+import pytest
+from .gateway_facade_input_validation import (
+    JSONParsingErrorException,
+    validate_deploy_input,
+    InvalidInputException,
+    get_abi_from_json,
+)
+
+COMPILED_CONTRACT_ABI_ONLY = """
```
Comment (kasperski95) [inline protostar/starknet_gateway/gateway_facade_input_validation_test.py:9]:
```suggestion
COMPILED_CONTRACT_ABI = """
```
How should `ONLY` be understood?

File: protostar/starknet_gateway/gateway_facade_input_validation.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
+import json
```
Comment (kasperski95) [inline protostar/starknet_gateway/gateway_facade_input_validation.py:1]: Does this file need to be prefixed with `gateway_facade`?  Maybe `input_validators.py` would be descriptive enough.

File: protostar/starknet_gateway/gateway_facade_input_validation.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+            "Couldn't parse given contract JSON.", str(ex)
+        ) from ex
+    except KeyError:
+        raise JSONParsingErrorException(
+            "No ABI found in the given compiled contract."
+        ) from KeyError
+    return abi
+
+
+def validate_deploy_input(compiled_contract_json: str, inputs: List[int]) -> None:
```
Comment (kasperski95) [inline protostar/starknet_gateway/gateway_facade_input_validation.py:48]:
```suggestion
def validate_contstructor_inputs(compiled_contract_json: str, inputs: List[int]) -> None:
```

This function doesn't need to be tied specifically to deploy.

File: protostar/starknet_gateway/gateway_facade_input_validation.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        abi = json.loads(compiled_contract_json)["abi"]
+    except json.decoder.JSONDecodeError as ex:
+        raise JSONParsingErrorException(
+            "Couldn't parse given contract JSON.", str(ex)
+        ) from ex
+    except KeyError:
+        raise JSONParsingErrorException(
+            "No ABI found in the given compiled contract."
+        ) from KeyError
+    return abi
```
Comment (kasperski95) [inline protostar/starknet_gateway/gateway_facade_input_validation.py:45]: Shouldn't this code be moved to `protostar.compiler` module?

This kind of code couples GatewayFacade with Starknet Internals. Consider following approach:

Component level: `protostar.starknet_gateway  <- protostar.compiler` 
Class level: `GatewayFacade  <- InputValidator  <= AbiInterpreter`

File: protostar/starknet_gateway/gateway_facade_input_validation.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+
+def validate_deploy_input(compiled_contract_json: str, inputs: List[int]) -> None:
+    abi = get_abi_from_json(compiled_contract_json)
+    constructor = get_constructor(abi)
```
Comment (MaksymilianDemitraszek) [inline protostar/starknet_gateway/gateway_facade_input_validation.py:56]:
```suggestion
    constructor_abi = get_constructor_abi(abi)
```

File: protostar/starknet_gateway/gateway_facade_input_validation.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+def get_constructor(abi: AbiType) -> ConstructorType:
+    candidates = [x for x in abi if x["type"] == "constructor"]
+    if not candidates:
+        return {"inputs": [], "outputs": []}
+    assert len(candidates) == 1
+    return candidates[0]
+
+
+def get_abi_from_json(compiled_contract_json: str) -> AbiType:
```
Comment (MaksymilianDemitraszek) [inline protostar/starknet_gateway/gateway_facade_input_validation.py:40]: Can we make separate module abi_utils for this? Doesn't such utils exist already?

File: protostar/starknet_gateway/gateway_facade_input_validation.py [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+
+def validate_deploy_input(compiled_contract_json: str, inputs: List[int]) -> None:
+    abi = get_abi_from_json(compiled_contract_json)
+    constructor = get_constructor(abi)
+    expected_inputs = constructor["inputs"]
+
+    serializer = CairoSerializer(identifier_manager_from_abi(abi))
+
+    serializer.to_python()  ## add pr to starknet.py
```
Comment (MaksymilianDemitraszek) [inline protostar/starknet_gateway/gateway_facade_input_validation.py:61]:
```suggestion
    serializer.to_python()  ## add pr to starknet.py
```
What add pr to starknet.py mean? Can you elaborate in comment/link issue

File: protostar/starknet_gateway/gateway_facade_input_validation.py [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+        try:
+            _, inputs = serializer.resolve_type(cairo_type).to_python(
+                cairo_type, name, inputs
+            )
+        except ValueError:
+            raise InvalidInputException("Not enough constructor arguments provided.")
+    if len(inputs) > 0:
+        raise InvalidInputException("Too many constructor arguments provided.")
+
+    return
```
Comment (MaksymilianDemitraszek) [inline protostar/starknet_gateway/gateway_facade_input_validation.py:74]: ?

File: protostar/starknet_gateway/gateway_facade_input_validation.py [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+class InvalidInputException(ProtostarException):
+    pass
+
+
+class JSONParsingErrorException(ProtostarException):
+    pass
+
+
+AbiType = List[Dict[str, Any]]
```
Comment (MaksymilianDemitraszek) [inline protostar/starknet_gateway/gateway_facade_input_validation.py:21]: Doesn't this alias already exist?

File: protostar/starknet_gateway/gateway_facade_input_validation.py [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+ConstructorType = Dict[str, Any]
```
Comment (MaksymilianDemitraszek) [inline protostar/starknet_gateway/gateway_facade_input_validation.py:22]: ConstructorAbiType?

File: protostar/starknet_gateway/gateway_facade_input_validation.py [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+        except ValueError:
+            raise InvalidInputException("Not enough constructor arguments provided.")
+    if len(inputs) > 0:
+        raise InvalidInputException("Too many constructor arguments provided.")
+
+    return
+
+    # print(serializer.from_python())
+
+    i = 0
```
Comment (MaksymilianDemitraszek) [inline protostar/starknet_gateway/gateway_facade_input_validation.py:78]: `accumulated_size`?

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]
